### PR TITLE
Improve Set marshaling to add encoding support, and decoding to map

### PR DESCRIPTION
**Pull request by kalman**
The only support that marshal has for Set at the moment is decoding to
slice.

### CODE REVIEW

**Comment by kalman** [comment]
For #2842.

File: go/marshal/decode.go [hunk 1]
```diff
@@ -386,32 +438,45 @@ func interfaceDecoder(t reflect.Type) decoderFunc {
...
 		return reflect.TypeOf(float64(0))
 	case types.StringKind:
 		return reflect.TypeOf("")
-	case types.ListKind, types.SetKind:
-		et := getGoTypeForNomsType(nt.Desc.(types.CompoundDesc).ElemTypes[0], rt, v)
-		return reflect.SliceOf(et)
+	case types.ListKind:
+		return listType()
+	case types.SetKind:
+		if shouldSetDecodeAsMap(rt, tags) {
```
Comment (kalman) [inline go/marshal/decode.go]: This change looked right, but now I realise it isn't. Reverting (just "getGoTypeForNomsType").

Comment (kalman) [inline go/marshal/decode.go]: Done, PTAL.

File: go/marshal/encode.go [hunk 2]
```diff
@@ -33,7 +33,8 @@ import (
 //
 // Slices and arrays are encoded as Noms types.List.
 //
-// Maps are encoded as Noms types.Map.
+// Maps are encoded as Noms types.Map, or a types.Set if the value type is
+// struct{} and the field is tagged with `noms:"key"`.
```
Comment (arv) [inline go/marshal/encode.go]: tagged with `noms:",set"`?

File: go/marshal/encode.go [hunk 3]
```diff
@@ -277,6 +282,10 @@ type encoderCacheT struct {
 
 var encoderCache = &encoderCacheT{}
 
+// Separate Set encoder cache because the same type with and without the
+// `noms:",set"` tag encode differently (Set vs Map).
+var setEncoderCache = &encoderCacheT{}
```
Comment (arv) [inline go/marshal/encode.go]: Maybe the cache could use a "tuple" as the key instead?

Comment (kalman) [inline go/marshal/encode.go]: The tuple would probably be (reflect.Type, types.NomsKind). I like the way it is now but if you prefer that I can change it.

File: go/marshal/encode.go (hunk 2 above)
Comment (kalman) [inline go/marshal/encode.go]: Done.

**Comment by kalman** [comment]
PTAL, test added.

and see response to comment.

**Comment by kalman** [comment]
THanks

**Pull request merged**